File: go-backend-interview-prep/04_interfaces_and_generics/02_interface_composition/main.py
# Python equivalent of Interface Composition (compare with main.go)
# Go interface embedding ~ Python multiple Protocol inheritance

from typing import Protocol


# --- Small, focused protocols ---
class Saver(Protocol):
    def save(self, data: str) -> None: ...

class Loader(Protocol):
    def load(self, id: str) -> str | None: ...

class Deleter(Protocol):
    def delete(self, id: str) -> None: ...


# --- Composed protocol (multiple inheritance) ---
class Storage(Saver, Loader, Deleter, Protocol):
    ...


# --- Concrete type satisfying all ---
class MemoryStore:
    def __init__(self):
        self.data: dict[str, str] = {}

    def save(self, data: str) -> None:
        key = f"key-{len(self.data) + 1}"
        self.data[key] = data
        print(f"  saved: {key} -> {data!r}")

    def load(self, id: str) -> str | None:
        return self.data.get(id)

    def delete(self, id: str) -> None:
        self.data.pop(id, None)
        print(f"  deleted: {id}")


def save_record(s: Saver, data: str) -> None:
    s.save(data)


def load_record(l: Loader, id: str) -> None:
    val = l.load(id)
    if val is None:
        print(f"  error: not found: {id}")
    else:
        print(f"  loaded: {id} -> {val!r}")


def main():
    store = MemoryStore()

    # --- Use as composed interface ---
    print("--- Full Storage interface ---")
    store.save("hello")
    store.save("world")

    # --- Use as Loader only ---
    print("\n--- As Loader only ---")
    load_record(store, "key-1")
    load_record(store, "key-99")

    # --- Use as Saver only ---
    print("\n--- As Saver only ---")
    save_record(store, "another")

    # --- Use as Deleter ---
    print("\n--- As Deleter ---")
    store.delete("key-1")

    print("\n--- Satisfies all sub-protocols ---")
    print("MemoryStore satisfies Saver, Loader, Deleter, and Storage")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/04_interfaces_and_generics/02_interface_composition/main.go
package main

import "fmt"

// --- Small, focused interfaces ---
type Saver interface {
	Save(data string) error
}

type Loader interface {
	Load(id string) (string, error)
}

type Deleter interface {
	Delete(id string) error
}

// --- Composed interface: embeds all three ---
type Storage interface {
	Saver
	Loader
	Deleter
}

// --- Concrete type satisfying Storage ---
type MemoryStore struct {
	data map[string]string
}

func NewMemoryStore() *MemoryStore {
	return &MemoryStore{data: make(map[string]string)}
}

func (m *MemoryStore) Save(data string) error {
	id := fmt.Sprintf("key-%d", len(m.data)+1)
	m.data[id] = data
	fmt.Printf("  saved: %s -> %q\n", id, data)
	return nil
}

func (m *MemoryStore) Load(id string) (string, error) {
	val, ok := m.data[id]
	if !ok {
		return "", fmt.Errorf("not found: %s", id)
	}
	return val, nil
}

func (m *MemoryStore) Delete(id string) error {
	delete(m.data, id)
	fmt.Printf("  deleted: %s\n", id)
	return nil
}

// --- Functions accepting sub-interfaces ---
func saveRecord(s Saver, data string) {
	s.Save(data)
}

func loadRecord(l Loader, id string) {
	val, err := l.Load(id)
	if err != nil {
		fmt.Println("  error:", err)
		return
	}
	fmt.Printf("  loaded: %s -> %q\n", id, val)
}

func main() {
	store := NewMemoryStore()

	// --- Use as composed interface ---
	fmt.Println("--- Full Storage interface ---")
	var s Storage = store
	s.Save("hello")
	s.Save("world")

	// --- Use as sub-interface ---
	fmt.Println("\n--- As Loader only ---")
	loadRecord(store, "key-1")
	loadRecord(store, "key-99")

	// --- Use as Saver only ---
	fmt.Println("\n--- As Saver only ---")
	saveRecord(store, "another")

	// --- Use as Deleter ---
	fmt.Println("\n--- As Deleter ---")
	store.Delete("key-1")

	// --- Pass where any sub-interface is expected ---
	// MemoryStore satisfies Saver, Loader, Deleter, AND Storage.
	fmt.Println("\n--- Satisfies all sub-interfaces ---")
	var saver Saver = store
	var loader Loader = store
	_ = saver
	_ = loader
	fmt.Println("MemoryStore satisfies Saver, Loader, Deleter, and Storage")
}
